src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java

```diff
@@ -0,0 +1,205 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(UidCoreClient.class);
+    private final IAttestationProvider attestationProvider;
+    private final ApplicationVersion appVersion;
+    private AtomicReference<String> attestationToken;
+    private Handler<Integer> responseWatcher;
+    private final String attestationEndpoint;
+    private final HttpClient httpClient;
+    private final IClock clock;
+    private ScheduledThreadPoolExecutor executor;
+    // Set this to be Instant.MAX so that if it's not set it won't trigger the re-attest
+    private Instant attestationTokenExpiresAt = Instant.MAX;
+    private final Lock lock;
+    private final AttestationTokenDecryptor attestationTokenDecryptor;
+
+
+    public AttestationTokenRetriever(String attestationEndpoint, ApplicationVersion appVersion, IAttestationProvider attestationProvider,
+                                     Handler<Integer> responseWatcher, IClock clock, HttpClient httpClient,
+                                     AttestationTokenDecryptor attestationTokenDecryptor, ScheduledThreadPoolExecutor executor) throws IOException {
+        this.attestationEndpoint = attestationEndpoint;
+        this.appVersion = appVersion;
+        this.attestationProvider = attestationProvider;
+        this.attestationToken = new AtomicReference<>(null);
+        this.responseWatcher = responseWatcher;
+        this.clock = clock;
+        this.lock = new ReentrantLock();
+        if (httpClient == null) {
+            this.httpClient = HttpClient.newHttpClient();
+        } else {
+            this.httpClient = httpClient;
+        }
+        if (attestationTokenDecryptor == null) {
+            this.attestationTokenDecryptor = new AttestationTokenDecryptor();
+        } else {
+            this.attestationTokenDecryptor = attestationTokenDecryptor;
+        }
+        if (executor == null) {
+            this.executor = new ScheduledThreadPoolExecutor(1);
+        } else {
+            this.executor = executor;
+        }
+    }
+
+    private void attestationExpirationCheck(){
+        Instant currentTime = clock.now();
+        Instant tenMinutesBeforeExpire = attestationTokenExpiresAt.minusSeconds(600);
+
+        stopAttestationExpirationCheck();
+
+        if (currentTime.isAfter(tenMinutesBeforeExpire)) {
+            LOGGER.info("Attestation token is 10 mins from the expiry timestamp %s. Re-attest...", attestationTokenExpiresAt);
+            try {
+                attestInternal();
+            }
+            catch (AttestationTokenRetrieverException | IOException e) {
+                notifyResponseStatusWatcher(401);
+                LOGGER.info("Re-attest failed: ", e.getMessage());
+            }
+        }
+    }
+
+    private void scheduleAttestationExpirationCheck() {
+        // Schedule the task to run every minute
+        executor.scheduleAtFixedRate(this::attestationExpirationCheck, 0, TimeUnit.MINUTES.toMillis(1), TimeUnit.MILLISECONDS);
```

Comment:
What happens if you try to call scheduleAtFixedRate if it is already scheduled?

Reply:
Added a check for that.